File: packages/veil-sdk-js/src/discovery_links.ts
export function buildVpsDiscoveryLink(options: {
  wsEndpoints?: string[];
  quicEndpoint?: string;
  quicCertHex?: string;
  quicCertB64?: string;
  peers?: string[];
  tags?: string[];
}): string {
  const params = new URLSearchParams();
  const ws = dedupe(options.wsEndpoints ?? []);
  for (const endpoint of ws) {
    if (endpoint) params.append("ws", endpoint);
  }
  if (options.quicEndpoint) {
    params.set("quic", options.quicEndpoint);
  }
  if (options.quicCertB64) {
    params.set("certb64", options.quicCertB64);
  } else if (options.quicCertHex) {
    params.set("cert", options.quicCertHex);
  }
  for (const peer of dedupe(options.peers ?? [])) {
    if (peer) params.append("peer", peer);
  }
  for (const tag of dedupe(options.tags ?? [])) {
    if (tag) params.append("tag", tag);
  }
  const query = params.toString();
  return query ? `veil://vps?${query}` : "veil://vps";
}

function dedupe(values: string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const value of values) {
    if (!seen.has(value)) {
      seen.add(value);
      out.push(value);
    }
  }
  return out;
}

File: packages/veil-sdk-dart/lib/src/discovery_links.dart
String buildVpsDiscoveryLink({
  List<String> wsEndpoints = const [],
  String? quicEndpoint,
  String? quicCertHex,
  String? quicCertB64,
  List<String> peers = const [],
  List<String> tags = const [],
}) {
  final params = <String, List<String>>{};
  final ws = _dedupe(wsEndpoints.where((e) => e.isNotEmpty));
  if (ws.isNotEmpty) {
    params["ws"] = ws;
  }
  if (quicEndpoint != null && quicEndpoint.isNotEmpty) {
    params["quic"] = [quicEndpoint];
  }
  final certB64 = quicCertB64?.trim() ?? "";
  final certHex = quicCertHex?.trim() ?? "";
  if (certB64.isNotEmpty) {
    params["certb64"] = [certB64];
  } else if (certHex.isNotEmpty) {
    params["cert"] = [certHex];
  }
  final peerList = _dedupe(peers.where((e) => e.isNotEmpty));
  if (peerList.isNotEmpty) {
    params["peer"] = peerList;
  }
  final tagList = _dedupe(tags.where((e) => e.isNotEmpty));
  if (tagList.isNotEmpty) {
    params["tag"] = tagList;
  }
  final query = params.entries
      .expand(
        (entry) => entry.value.map(
          (value) =>
              "${Uri.encodeQueryComponent(entry.key)}=${Uri.encodeQueryComponent(value)}",
        ),
      )
      .join("&");
  return query.isEmpty ? "veil://vps" : "veil://vps?$query";
}

List<String> _dedupe(Iterable<String> values) {
  final out = <String>[];
  final seen = <String>{};
  for (final value in values) {
    if (seen.add(value)) {
      out.add(value);
    }
  }
  return out;
}
